File: automl_server/app/utils/annotation_classes.py
"""Helpers for annotation class name parsing."""
import json
import re
from typing import Iterable, Optional


CLASS_DELIMITER_PATTERN = re.compile(r"[,;\n\r；，]+")


def _normalize_items(items: Iterable[object]) -> list[str]:
    seen: set[str] = set()
    classes: list[str] = []

    for item in items:
        class_name = str(item).strip()
        if not class_name or class_name in seen:
            continue
        seen.add(class_name)
        classes.append(class_name)

    return classes


def parse_annotation_classes(raw_classes: Optional[str]) -> list[str]:
    if not raw_classes:
        return []

    try:
        parsed = json.loads(raw_classes)
        if isinstance(parsed, list):
            return _normalize_items(parsed)
        if isinstance(parsed, str):
            return _normalize_items(CLASS_DELIMITER_PATTERN.split(parsed))
    except Exception:
        return _normalize_items(CLASS_DELIMITER_PATTERN.split(raw_classes))

    return []


def serialize_annotation_classes(raw_classes: Optional[str]) -> str:
    classes = parse_annotation_classes(raw_classes)
    return json.dumps(classes, ensure_ascii=False)

File: frontend_v2/src/utils/annotationClasses.ts
function normalizeClassItems(items: unknown[]): string[] {
  const seen = new Set<string>();
  const classes: string[] = [];

  for (const item of items) {
    const className = String(item).trim();
    if (!className || seen.has(className)) continue;
    seen.add(className);
    classes.push(className);
  }

  return classes;
}

function parseClassText(text: string): string[] {
  return normalizeClassItems(text.split(/[,\n\r;；，]+/));
}

export function parseAnnotationClasses(rawClasses: string | string[] | null | undefined): string[] {
  if (!rawClasses) return [];
  if (Array.isArray(rawClasses)) return normalizeClassItems(rawClasses);

  try {
    const parsed = JSON.parse(rawClasses);
    if (Array.isArray(parsed)) return normalizeClassItems(parsed);
    if (typeof parsed === 'string') return parseClassText(parsed);
  } catch {
    return parseClassText(rawClasses);
  }

  return [];
}

export function serializeAnnotationClasses(rawClasses: string | string[] | null | undefined): string {
  return JSON.stringify(parseAnnotationClasses(rawClasses));
}
